### DISCUSSION TITLE
Limiting multipart file size

### DISCUSSION THREAD

**Discussion by mqchen-ws** [resteasy/resteasy#4235] [Q&A]
Version: 6.2.9.Final

Hi,

I'm trying to limit the maximum file size buffered onto disk as a part of a multipart upload. The code is pretty simple:

```
    @POST
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    public void upload(final MultipartInput input) {
      // etc.
    }
```

but I can't seem to set an upper bound through any properties, etc. After using a debugger and stepping through some of the code, I found this:

[Mime4JWorkaround#parseMessage](https://github.com/resteasy/resteasy/blob/main/providers/multipart/src/main/java/org/jboss/resteasy/plugins/providers/multipart/Mime4JWorkaround.java#L61)

which seems to be taking in the HttpInput from the request and creating `BodyPart` backed by temp files. The `MimeConfig` config has a `maxContentLen` which seems to do what I want, but that workaround class is using the default config with -1 (unlimited).

Is there another way to limit the upload file size? Or perhaps a way to use a custom value for that config?

Thank you!

**Comment by jamezp** [comment]
I don't see a way at first glance to limit the size on disk. I'll take a closer look and write a test for it however. Then, assuming a deeper look doesn't turn up anything, I'll file a JIRA and report back here either way.

**Comment by mqchen-ws** [comment]
Alternatively, another approach that would give a similar result is handling the underlying multipart input stream directly instead of buffering, but I didn't see a way to do that either, since `MultipartInput` already gives a list of processed `InputPart` after in memory or disk buffering has taken place.

A coworker looked around and found a way to directly use `MimeStreamParser` and writing our own `ContentHandler` to process the http input stream directly, but it would be nice to be able to have it supported directly, if possible.

perhaps something like this?
```
public interface StreamedMultipartInput {

    Stream<InputPart> getPartsStream();

    String getPreamble();

    void close();
}
```

**Reply by liweinan** [reply]
Hi @mqchen-ws It would be great if you can provide a pull request to what you and your colleague have made this work, or providing an example on the way you solved this problem. Which will be very helpful to others that have similar requirements.

**Reply by mqchen-ws** [reply]
Something like this. You have to keep a bit of state yourself in the handler to track what needs to be processed next.

```    
    @POST
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    public void upload(
            @HeaderParam(HttpHeaders.CONTENT_TYPE) final String contentType,
            @Context final HttpServletRequest req) {
        final MimeConfig config = MimeConfig
                .custom()
                .setMaxContentLen(50 * 1024 * 1024)
                .setHeadlessParsing(contentType)
                .build();
        final MimeTokenStream tokenStream =
                new MimeTokenStream(config, null, null, new DefaultBodyDescriptorBuilder());
        final MimeStreamParser parser = new MimeStreamParser(tokenStream);
        final ContentHandler handler = new AbstractContentHandler() {
            @Override
            public void field(final Field field) {
                // handle field, e.g. read as ContentDispositionField to get file name of multipart file
            }

            @Override
            public void body(final BodyDescriptor bd, final InputStream is) throws IOException {
               // handle body
            }
        }
        parser.setContentHandler(handler);
        try (InputStream input = req.getInputStream()) {
            parser.parse(input);
        }
    }
```

**Comment by jamezp** [comment]
@mqchen-ws It looks like you can limit the upload size https://docs.jboss.org/resteasy/docs/6.2.9.Final/userguide/#_entity_part. The two properties you'd be interested in are:

- `dev.resteasy.entity.memory.threshold`
- `dev.resteasy.entity.file.threshold`

Sorry for the delay on this. I knew the properties worked if the `EntityPart` was used, I just wasn't sure if the legacy RESTEasy API worked.
